instrumentation/servlet/servlet-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/snippet/OutputStreamSnippetInjectionHelper.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.servlet.snippet;
+
+import static java.util.logging.Level.FINE;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+
+public class OutputStreamSnippetInjectionHelper {
+
+  private static final Logger logger =
+      Logger.getLogger(OutputStreamSnippetInjectionHelper.class.getName());
+
+  private final Supplier<String> snippetSupplier;
+
+  public OutputStreamSnippetInjectionHelper(Supplier<String> snippetSupplier) {
+    this.snippetSupplier = snippetSupplier;
+  }
+
+  public String getSnippet() {
+    return snippetSupplier.get();
+  }
+
+  /**
+   * return true means this method performed the injection, return false means it didn't inject
+   * anything Servlet3OutputStreamWriteAdvice would skip the write method when the return value is
+   * true, and would write the original bytes when the return value is false.
+   */
+  public boolean handleWrite(
+          InjectionState state, OutputStream out, byte[] original, int off, int length)
+      throws IOException {
+    if (state.isHeadTagWritten()) {
+      return false;
+    }
+    int endOfHeadTagPosition;
+    boolean endOfHeadTagFound = false;
+    for (endOfHeadTagPosition = off;
+        endOfHeadTagPosition < length && endOfHeadTagPosition - off < length;
+        endOfHeadTagPosition++) {
+      if (state.processByte(original[endOfHeadTagPosition])) {
+        endOfHeadTagFound = true;
+        break;
+      }
+    }
+    if (!endOfHeadTagFound) {
+      return false;
+    }
+
+    if (state.getWrapper().isNotSafeToInject()) {
+      return false;
+    }
+    byte[] snippetBytes;
+    try {
+      snippetBytes = snippetSupplier.get().getBytes(state.getCharacterEncoding());
+    } catch (UnsupportedEncodingException e) {
+      logger.log(FINE, "UnsupportedEncodingException", e);
```

Comment:
`UnsupportedEncodingException` as exception message is redundant as it would be printed as part of the exception stack trace anyway. Usually we'd use something that describes the failed action as the message like `Failed getting snippet bytes`.